### PR TITLE
Fix panic in mprotect when merging adjacent mappings

### LINKED ISSUE #2917: Reachable unwrap panic in vmar protect when merging mappings
### Describe the bug
`inner.remove(&vm_mapping_range.start).unwrap()` in `kernel/src/vm/vmar/vmar_impls/protect.rs` panics when `inner.remove` returns `None`.

This happens when `mprotect` is called on a range covering multiple adjacent mappings. If a prior mapping in the range is updated and then merged with the next mapping (because permissions now match), the next mapping is removed from `inner.vm_mappings` during the merge. However, the loop continues to iterate over the list of mappings found at the start of the function, and tries to process this "next" mapping in the subsequent iteration. It attempts to remove it from the tree again, leading to a panic.

https://github.com/asterinas/asterinas/blob/b7ae7383102800529a079f00f92178511d35b8f3/kernel/src/vm/vmar/vmar_impls/protect.rs#L46

### To Reproduce
```C
#define _GNU_SOURCE
#include <sys/mman.h>
#include <unistd.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#define PAGE_SIZE 4096
#ifndef MAY_READ
#define MAY_READ 0x8
#endif

int main() {
    // 1. Map two adjacent pages.
    // Initial state: One merged mapping covering 2 pages with RW permissions.
    void *p = mmap(NULL, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE, 
                   MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (p == MAP_FAILED) {
        perror("mmap");
        return 1;
    }
    
    printf("Mapped at %p\n", p);

    // 2. Change permissions of the first page to split the mapping.
    // Result:
    // M1 (Page 0): PROT_READ
    // M2 (Page 1): PROT_READ | PROT_WRITE
    if (mprotect(p, PAGE_SIZE, PROT_READ) < 0) {
        perror("mprotect split");
        return 1;
    }
    printf("Split done\n");

    // 3. Trigger the panic.
    // Call mprotect on the whole range with PROT_READ | PROT_WRITE | MAY_READ (0x8).
    // The specific flag combination isn't strictly necessary, just needs to trigger an update + merge.
    
    printf("Triggering panic...\n");
    if (mprotect(p, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE | MAY_READ) < 0) {
        perror("mprotect trigger");
        return 1;
    }

    printf("Did not panic?\n");
    return 0;
}
```

### Environment
- Asterinas version: main b7ae7383102800529a079f00f92178511d35b8f3
- Official docker asterinas/asterinas:0.16.2-20251211

### Logs
```
~ # /etc/repro_unwrap_panic.c 
Mapped at 0x7ffffeff8000
Split done
Triggering panic...
[    24.491] ERROR: Uncaught panic:
        called `Option::unwrap()` on a `None` value
        at /root/asterinas/kernel/src/vm/vmar/vmar_impls/protect.rs:46
        on CPU 0 by thread Some(Thread { task: (Weak), data: Any { .. }, is_exited: false, cpu_affinity: AtomicIdSet { bits: [1], phantom: PhantomData<ostd::cpu::id::CpuId> }, sched_attr: SchedAttr { policy: SchedPolicyState { kind: AtomicSchedPolicyKind(2), policy: UnsafeCell { .. } }, last_cpu: AtomicCpuId(0), real_time: RealTimeAttr { prio: 99, time_slice: 31489960 }, fair: FairAttr { weight: 1024, pending_weight: 0, vruntime: 290882507 } } })
Printing stack trace:
   1: fn 0xffffffff889fbae0 - pc 0xffffffff889fbb2b / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70260;

   2: fn 0xffffffff8804cbf0 - pc 0xffffffff8804d3ed / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70280;

   3: fn 0xffffffff8804e740 - pc 0xffffffff8804e74a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70650;

   4: fn 0xffffffff8804b000 - pc 0xffffffff8804b00a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70660;

   5: fn 0xffffffff88aec890 - pc 0xffffffff88aec8d6 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70670;

   6: fn 0xffffffff88aec860 - pc 0xffffffff88aec886 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd706b0;

   7: fn 0xffffffff88ad29e0 - pc 0xffffffff88ad29f5 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd706d0;

   8: fn 0xffffffff88656700 - pc 0xffffffff88656c4a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd706e0;

   9: fn 0xffffffff8853ef80 - pc 0xffffffff8853f33a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70db0;

  10: fn 0xffffffff88577640 - pc 0xffffffff8857c65b / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd70fa0;

  11: fn 0xffffffff88574bc0 - pc 0xffffffff88574c5a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd844f0;

  12: fn 0xffffffff884cdcc0 - pc 0xffffffff884ce571 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84650;

  13: fn 0xffffffff884ce730 - pc 0xffffffff884ce78e / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84b40;

  14: fn 0xffffffff883ee270 - pc 0xffffffff883ee29e / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84cf0;

  15: fn 0xffffffff883ee1a0 - pc 0xffffffff883ee1e6 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84d20;

  16: fn 0xffffffff88436a60 - pc 0xffffffff88436a8c / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84d90;

  17: fn 0xffffffff886c7010 - pc 0xffffffff886c704f / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84dd0;

  18: fn 0xffffffff884ce6f0 - pc 0xffffffff884ce717 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84f40;

  19: fn 0xffffffff88106e70 - pc 0xffffffff88106e8d / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84f80;

  20: fn 0xffffffff889e5c30 - pc 0xffffffff889e5c55 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84fa0;

  21: fn 0xffffffff88960800 - pc 0xffffffff88960884 / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd84fd0;

  22: fn 0xffffffff889e8a31 - pc 0xffffffff889e8a3a / registers:

     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd85000;


     rax 0xffffdfffffd70178; rdx 0xffffffff8894f462; rcx  0x100000201002693; rbx                0x0;
     rsi                0x0; rdi 0xffffffff88bdfd58; rbp                0x0; rsp 0xffffdfffffd85008;

OSDK] The kernel seems panicked. Parsing stack trace for source lines:
(  1) ostd::panic::print_stack_trace
          at /root/asterinas/ostd/src/panic.rs:113
(  2) aster_kernel::thread::oops::panic_handler
          at /root/asterinas/kernel/src/thread/oops.rs:130
(  3) __ostd_panic_handler
          at ??:?
(  4) __rustc::rust_begin_unwind
          at 7nscgio2smpzudpoic612j35d:?
(  5) core::panicking::panic_fmt
          at ??:?
(  6) core::panicking::panic
          at ??:?
(  7) core::option::unwrap_failed
          at ??:?
(  8) <core::option::Option<aster_kernel::vm::vmar::vm_mapping::VmMapping>>::unwrap
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1015
(  9) aster_kernel::syscall::mprotect::sys_mprotect
          at /root/asterinas/kernel/src/syscall/mprotect.rs:48
( 10) aster_kernel::syscall::arch::syscall_dispatch
          at /root/asterinas/kernel/src/syscall/mod.rs:218
( 11) aster_kernel::syscall::handle_syscall
          at /root/asterinas/kernel/src/syscall/mod.rs:375
( 12) aster_kernel::thread::task::create_new_user_task::{closure#0}
          at /root/asterinas/kernel/src/thread/task.rs:95
( 13) aster_kernel::thread::task::create_new_user_task::{closure#1}
          at /root/asterinas/kernel/src/thread/task.rs:126
( 14) unwinding::panicking::catch_unwind::do_call::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:56
( 15) unwinding::panicking::catch_unwind::<unwinding::panic::RustPanic, (), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panicking.rs:42
( 16) unwinding::panic::catch_unwind::<(), aster_kernel::thread::task::create_new_user_task::{closure#1}>
          at /root/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/unwinding-0.2.8/src/panic.rs:87
( 17) aster_kernel::thread::oops::catch_panics_as_oops::<aster_kernel::thread::task::create_new_user_task::{closure#1}, ()>
          at /root/asterinas/kernel/src/thread/oops.rs:55
( 18) aster_kernel::thread::task::create_new_user_task::{closure#2}
          at /root/asterinas/kernel/src/thread/task.rs:131
( 19) <aster_kernel::thread::task::create_new_user_task::{closure#2} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250
( 20) <alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output = ()> + core::marker::Send> as core::ops::function::FnOnce<()>>::call_once
          at /root/.rustup/toolchains/nightly-2025-12-06-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2206
( 21) kernel_task_entry
          at /root/asterinas/ostd/src/task/mod.rs:196
( 22) kernel_task_entry_wrapper
          at ??:?
make: *** [Makefile:308: run_kernel] Error 1
```

### PR DESCRIPTION
**Pull request by Marsman1996**
Fix #2917

### CODE REVIEW

**Comment by Marsman1996** [comment]
Well, did not notice there is already a PR #2918  ...

**Comment by lrh2000** [comment]
*If this is correct*, there should be comments before `continue`:
```rust
// This can happen only if the mapping is merged to the previous one (just
// protected before). We can skip this mapping because its property is already
// correct.
```

**Comment by Marsman1996** [comment]
> _If this is correct_, there should be comments before `continue`:
> 
> ```rust
> // This can happen only if the mapping is merged to the previous one (just
> // protected before). We can skip this mapping because its property is already
> // correct.
> ```

I'm not 100% sure about this, since the logic here is different from Linux🤔

**Comment by Marsman1996** [comment]
But CortenMM will change everything here. 
When will that PR be merged? Maybe it will solve the problem.

**Comment by junyang-zh** [comment]
> But CortenMM will change everything here. When will that PR be merged? Maybe it will solve the problem.

I tested your program on #2615, and the kernel runs well.

Perhaps there are still some days till the merge of CortenMM. We can merge this PR first.

**Comment by lrh2000** [comment]
Regardless of whether or not CortenMM changes this, we still want the change to be correct, don't we?

Please add the comments here as well. This is not related to whether the code will be changed later...

**Comment by tatetian** [comment]
@Marsman1996 Could you add a regression test that reproduces #2917 under [`test/initramfs/apps/mmap`](https://github.com/asterinas/asterinas/tree/main/test/initramfs/src/apps/mmap)? This helps ensure that even after CortenMM refactors the code, this edge case is still handled properly.

**Comment by Marsman1996** [comment]
Regression test  added

**Comment by lrh2000** [comment]
> Regression test added

Could you please also address https://github.com/asterinas/asterinas/pull/2926#discussion_r2726465132 in the newly added regression test? Specially:
>> * Add the cleanup `munmap()`s. (Also in the newly added test `mprotect_permission`):

In addition, I think we should add blank lines after `mmap` and before `munmap`, just like your previously added test `mprotect_invalid_permission`.

File: test/initramfs/src/apps/mmap/mmap_and_mprotect.c [hunk 1]
```diff
@@ -69,3 +69,17 @@ FN_TEST(mprotect_invalid_permission)
...
+
+FN_TEST(mprotect_merge_after_split_mapping)
+{
+	void *addr = TEST_SUCC(mmap(NULL, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE,
+				    MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+
+	// Split the mapping into two.
+	TEST_SUCC(mprotect(addr, PAGE_SIZE, PROT_READ));
+	// After the first mapping is protected, the two mappings will merge.
+	TEST_SUCC(mprotect(addr, 2 * PAGE_SIZE, PROT_READ | PROT_WRITE));
```
Comment (hburaylee) [inline test/initramfs/src/apps/mmap/mmap_and_mprotect.c]: This test case can also pass without the fix from the patch above, because during the second iteration of the loop, it will enter the `continue` condition:
```rust
for (vm_mapping_range, vm_mapping_perms) in protect_mappings {
  ...
  if perms == vm_mapping_perms & VmPerms::ALL_PERMS {
     continue;
  }
  ...
}
```

**Review by junyang-zh — APPROVED** [review]
LGTM!

**Comment by tatetian** [comment]
@Marsman1996 Sorry for the delay on merging this PR. Could you rebase the fix?

**Comment by Marsman1996** [comment]
> Could you rebase the fix?

Rebased
